### ISSUE TITLE
TPM Software stack in OP-TEE

### ISSUE THREAD

**Issue by dmcilvaney** [OP-TEE/optee_os#3219]
### Related:
#1867
#3189

As per #3189 we are actively working on attestation and code measurement during boot. Something we are now considering is offloading this behavior into a TPM (discreet, or possibly firmware as a fallback).

**I know there has been lots of activity around the fTPM, but has anyone looked into adding TPM support to OP-TEE itself? If so, is there a preferred TPM stack?** We are looking at prototyping this with https://github.com/tpm2-software/tpm2-tss, maybe as a PTA.

Newer TPMs (as of v.138) support key derivation based on an external measurement (salt) + a secret seed value inside the TPM, which itself can be from an earlier salted key (#3189 explains a very similar flow as implemented by the firmware itself). It would be nice to have a single point of contact (the TPM) for measurement/attestation at every level of boot firmware instead of having to customize code for every firmware component. 

We can also offload the policy for TAs to the TPM, so that OP-TEE no longer needs  a hardcoded key to decide if it can run a TA or not. As another bonus, things like RNG, key generation, etc. can be offloaded to a very secure hardware device.

Obvious downsides include adding a dependency on hardware SPI or similar, adding the complexity of the TPM software stack to OP-TEE, and if there is a desire to use the firmware TPM, the chicken/egg question of how to access secure storage if we are using the fTPM to generate the keys to unlock RPMB (seems like it should be solvable though).

I'm curious if anyone has done any work towards this goal. This is something that is very important for shipping enterprise/government grade security.

## tldr:
**1. Has anyone looked into accessing a TPM from inside OP-TEE before?**
**2. If so, is there a preference for how the TPM should interface with OP-TEE?**

Thanks,
Daniel

**Comment by bzhu5** [comment]
> Obvious downsides include adding a dependency on hardware SPI or similar, adding the complexity of the TPM software stack to OP-TEE

May not need to add full tss stack as I think op-tee only uses very small subsets of discrete tpm functionalities.

**Comment by jbech-linaro** [comment]
Hi Daniel,

> **1. Has anyone looked into accessing a TPM from inside OP-TEE before?**

Not what I'm aware of. So far we've only looked at implementing it in firmware as you've already seen.

> **2. If so, is there a preference for how the TPM should interface with OP-TEE?**

That's an interesting question with many answers I suppose. If you would have made the TPM available only to the TEE, then I guess you could ignore the upper levels in the TCG software stack and probably just implement the lower levels. I.e., ignore FAPI and eventually ESAPI. As you already mentioned, the communication will most likely be with SPI, so a PTA will be needed for the communication as such if nothing else.

On the other hand, if the TPM shall be accessible both from Linux and the TEE, then things start to become a bit challenging, since now you have to pay attention to the resource manager, which runs on the non-secure side. One could of course do something similar in tee-supplicant as we are doing for RPMB (piggy back on kernel services / APIs), but then you loose lots of the benefits by using the TPM from the TA (TEE) itself.

What is your own view on this besides that initial comment in this thread?

The comment from @bzhu5 shouldn't be ignored. Doing full TPM if you only need a small subset might be a bit overkill.

**Comment by jforissier** [comment]
> On the other hand, if the TPM shall be accessible both from Linux and the TEE, then things start to become a bit challenging, since now you have to pay attention to the resource manager, which runs on the non-secure side. 

Perhaps the resource manager could run on the secure side instead and Linux would call into a PTA to obtain TPM services? Leveraging the TEE bus driver [1] contributed by @b49020.

[1] https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?h=v5.4-rc2&id=0fc1db9d1059

**Comment by dmcilvaney** [comment]
I'm not super familiar with TPMs, so I'll try and get our local expert here for some perspective. That said, some thoughts based on my best understanding:

From a high level, the operations we need involve measuring binaries, generating keys (one set of keys for each TA, derived from their binary measurements), and signing certificates. Each TA would get an x509 certificate chain rooted down to some hardware root of trust.

We have looked into implementing a VERY abbreviated stack in SPL (a fill-in-the-blanks set of pre-made buffers we send over SPI). We could do the same in OP-TEE reasonably easily, but the compactness and simplicity means it is very rigid and non-extensible.

As for the resource manager, I'm not familiar with the Linux drivers. Do you know if it is able to restart jobs which are pre-empted by a higher priority job? For the various tasks we have in mind we would need to load different keys in every time so we don't need to worry about state from the TEE side. If the Linux side can recover from an interrupted job then we can just ignore it. The OP-TEE tasks would generally only run when a new TA is loaded for the first time so there shouldn't be a big hit to performance.
